content/post/22-05-25-progress-report-34.md

```diff
@@ -0,0 +1,458 @@
+
+Here's what the `showgridnodes` debug command looks like:
+
+{{< imgw "images/post/pr_34/nodes" >}}
+
+{{< newsection >}}
+### Custom MIDI Soundfont Support
+*Contributed by vulppine*
+
+MIDI sound fonts are now selectable by the end user in Robust Toolbox! They override any fonts loaded in by the server, and the default sound font used by your current system. This is beneficial for when you want to use other sound fonts in SS14 rather than the default one installed on your system, without having to replace any system files or deal with any configuration. In order to add any custom sound fonts to any Robust Toolbox game, you'll need to go into the user configuration folder related to the game in question, and add any sound fonts you want into the 'soundfonts' folder.
+
+{{< newsection >}}
+### Replaced Lidgren Encryption with Libsodium
+*Contributed by PJB*
+
+Since we added account authentication, communication with the game server has mostly been AES encrypted. This was built using Lidgren's `NetEncryption` system and, well, it worked. If you [check the code](https://github.com/space-wizards/SpaceWizards.Lidgren.Network/blob/11198c1c46ea721dbcbe99296570495ba9cd68a6/Lidgren.Network/Encryption/NetCryptoProviderBase.cs) and you know a little bit about AES however... Yeah it reuses IVs on CBC mode. Yeah that makes it automatically broken. Hooray!
+
+Because I am not a cryptographic expert I decided that the best way to fix this was to simply replace this with Libsodium's [AEAD XChaCha20-Poly1305 API](https://doc.libsodium.org/secret-key_cryptography/aead/chacha20-poly1305/xchacha20-poly1305_construction). I trust that the crypto is not completely broken now. I also completely tore out Lidgren's encryption support in our fork (it was evidently broken and didn't integrate with the internals anyways).
+
+There are some gotchas here though. The first thing is that .NET didn't have an existing non-terrible Libsodium binding, so I had to make a partial binding for the APIs I care about myself: [SpaceWizards.Sodium](https://github.com/space-wizards/SpaceWizards.Sodium). 
+
+The second problem is that this change does have a noticeable increase on bandwidth usage. This is because proper encryption:tm: has more overhead in the form of AEAD and IV that has to be sent. Lidgren has a feature where it combines many net messages we send into fewer larger UDP packets. This is good, but the missed optimization here is that encryption runs *before* this combining. Ideally we would run encryption afterwards and only needing to encrypt single UDP packets, saving us some bandwidth. This would require deeper integration with Lidgren's internals however, so this is something for the future.
```

Comment:
```suggestion
The second problem is that this change does have a noticeable increase on bandwidth usage. This is because proper encryption™ has more overhead in the form of AEAD and IV that has to be sent. Lidgren has a feature where it combines many net messages we send into fewer larger UDP packets. This is good, but the missed optimization here is that encryption runs *before* this combining. Ideally we would run encryption afterwards and only needing to encrypt single UDP packets, saving us some bandwidth. This would require deeper integration with Lidgren's internals however, so this is something for the future.
```